File: android/app/src/main/kotlin/io/mynote/app/storage/GoogleDriveFolder.kt
package io.mynote.app.storage

import io.mynote.core.MyNoteJson
import io.mynote.core.RemoteFile
import io.mynote.core.RemoteFolder
import io.mynote.core.RemoteFolderError
import io.mynote.core.RemoteFolderException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.HttpUrl.Companion.toHttpUrl
import java.io.IOException

/**
 * Backup into a "MyNote" folder in the user's Google Drive.
 *
 * The cross-platform option: the same folder is read by the iOS app, so a note
 * written on an Android phone shows up on an iPhone. The file format is defined
 * once in `:core` and shared by both.
 */
class GoogleDriveFolder(
    private val auth: GoogleDriveAuth,
    private val client: OkHttpClient = OkHttpClient(),
) : RemoteFolder {

    override val displayName = "Google Drive"

    private var folderId: String? = null

    /** Drive addresses files by opaque id, not by name. */
    private val fileIds = mutableMapOf<String, String>()

    override suspend fun list(): List<RemoteFile> = withContext(Dispatchers.IO) {
        val folder = folder()
        val items = mutableListOf<RemoteFile>()
        var pageToken: String? = null

        do {
            val url = "$API/files".toHttpUrl().newBuilder()
                .addQueryParameter("q", "'$folder' in parents and trashed=false")
                .addQueryParameter("fields", "nextPageToken,files(id,name,modifiedTime,size,version)")
                .addQueryParameter("pageSize", "200")
                .apply { pageToken?.let { addQueryParameter("pageToken", it) } }
                .build()

            val listing = MyNoteJson.decodeFromString(
                FileList.serializer(),
                send(Request.Builder().url(url).get()),
            )
            for (file in listing.files) {
                fileIds[file.name] = file.id
                items += RemoteFile(
                    name = file.name,
                    modifiedAt = 0L,
                    size = file.size?.toLongOrNull() ?: 0L,
                    // Drive bumps `version` on every content or metadata change,
                    // which is exactly the "has this changed" signal we need.
                    version = file.version ?: file.modifiedTime ?: "0",
                )
            }
            pageToken = listing.nextPageToken
        } while (pageToken != null)

        items
    }

    override suspend fun read(name: String): ByteArray = withContext(Dispatchers.IO) {
        val id = id(name) ?: throw RemoteFolderException(RemoteFolderError.NotFound(name))
        sendBytes(Request.Builder().url("$API/files/$id?alt=media").get())
    }

    override suspend fun write(name: String, data: ByteArray) = withContext(Dispatchers.IO) {
        val existing = id(name)
        if (existing != null) {
            // Replace the contents, keeping the same id so other devices' notion
            // of the file and Drive's revision history stay intact.
            send(
                Request.Builder()
                    .url("$UPLOAD/files/$existing?uploadType=media")
                    .patch(data.toRequestBody(JSON))
            )
            return@withContext
        }

        // New file: metadata and content in one multipart request.
        val boundary = "mynote-${System.nanoTime()}"
        val metadata = MyNoteJson.encodeToString(
            NewFile.serializer(),
            NewFile(name = name, parents = listOf(folder())),
        )
        val body = buildString {
            append("--$boundary\r\nContent-Type: application/json; charset=UTF-8\r\n\r\n")
            append(metadata)
            append("\r\n--$boundary\r\nContent-Type: application/json\r\n\r\n")
        }.toByteArray() + data + "\r\n--$boundary--\r\n".toByteArray()

        val created = MyNoteJson.decodeFromString(
            DriveFile.serializer(),
            send(
                Request.Builder()
                    .url("$UPLOAD/files?uploadType=multipart&fields=id")
                    .post(body.toRequestBody("multipart/related; boundary=$boundary".toMediaType()))
            ),
        )
        fileIds[name] = created.id
    }

    override suspend fun delete(name: String) {
        withContext(Dispatchers.IO) {
            val id = id(name) ?: return@withContext
            send(Request.Builder().url("$API/files/$id").delete())
            fileIds.remove(name)
        }
    }

    // MARK: - Folder resolution

    /** Find the MyNote folder, creating it on first use. */
    private suspend fun folder(): String {
        folderId?.let { return it }

        val url = "$API/files".toHttpUrl().newBuilder()
            .addQueryParameter("q", "mimeType='$FOLDER_MIME' and name='$FOLDER_NAME' and trashed=false")
            .addQueryParameter("fields", "files(id,name)")
            .build()
        val existing = MyNoteJson.decodeFromString(
            FileList.serializer(),
            send(Request.Builder().url(url).get()),
        )
        existing.files.firstOrNull()?.let {
            folderId = it.id
            return it.id
        }

        val created = MyNoteJson.decodeFromString(
            DriveFile.serializer(),
            send(
                Request.Builder()
                    .url("$API/files?fields=id")
                    .post(
                        MyNoteJson.encodeToString(
                            NewFile.serializer(),
                            NewFile(name = FOLDER_NAME, parents = null, mimeType = FOLDER_MIME),
                        ).toRequestBody(JSON)
                    )
            ),
        )
        folderId = created.id
        return created.id
    }

    private suspend fun id(name: String): String? {
        fileIds[name]?.let { return it }
        list()
        return fileIds[name]
    }

    // MARK: - Transport

    private suspend fun send(builder: Request.Builder): String =
        sendBytes(builder).decodeToString()

    private suspend fun sendBytes(builder: Request.Builder): ByteArray {
        val request = builder
            .header("Authorization", "Bearer ${auth.token()}")
            .build()

        val response = try {
            client.newCall(request).execute()
        } catch (e: IOException) {
            throw RemoteFolderException(RemoteFolderError.Offline)
        }

        response.use {
            val bytes = it.body?.bytes() ?: ByteArray(0)
            if (!it.isSuccessful) throw RemoteFolderException(error(it.code, bytes.decodeToString()))
            return bytes
        }
    }

    private fun error(status: Int, body: String): RemoteFolderError = when {
        status == 401 -> RemoteFolderError.NeedsReauthentication
        status == 403 && body.contains("insufficient", true) -> RemoteFolderError.NeedsReauthentication
        status == 403 && body.contains("quota", true) -> RemoteFolderError.StorageFull
        status == 404 -> RemoteFolderError.NotFound(body)
        // Rate limited or Drive is unwell; both are worth retrying later.
        status == 429 || status >= 500 -> RemoteFolderError.Offline
        else -> RemoteFolderError.Provider("Drive error ($status)")
    }

    // MARK: - Wire types

    @Serializable
    private data class FileList(
        val files: List<DriveFile> = emptyList(),
        val nextPageToken: String? = null,
    )

    @Serializable
    private data class DriveFile(
        val id: String,
        val name: String = "",
        val modifiedTime: String? = null,
        val size: String? = null,
        val version: String? = null,
    )

    @Serializable
    private data class NewFile(
        val name: String,
        val parents: List<String>? = null,
        val mimeType: String? = null,
    )

    private companion object {
        const val FOLDER_NAME = "MyNote"
        const val FOLDER_MIME = "application/vnd.google-apps.folder"
        const val API = "https://www.googleapis.com/drive/v3"
        const val UPLOAD = "https://www.googleapis.com/upload/drive/v3"
        val JSON = "application/json; charset=utf-8".toMediaType()
    }
}

File: ios/MyNote/Sources/Storage/GoogleDriveFolder.swift
import Foundation
import MyNoteCore

/// Backup into a "MyNote" folder in the user's Google Drive.
///
/// The cross-platform option: the same folder is readable by the Android app, so
/// a note written on an iPhone shows up on an Android tablet. iCloud Drive
/// cannot do this — Apple publishes no Android API for it.
actor GoogleDriveFolder: RemoteFolder {
    nonisolated let displayName = "Google Drive"

    private static let folderName = "MyNote"
    private static let folderMime = "application/vnd.google-apps.folder"
    private static let api = "https://www.googleapis.com/drive/v3"
    private static let upload = "https://www.googleapis.com/upload/drive/v3"

    private let auth: GoogleAuth
    private var folderId: String?
    /// Drive addresses files by opaque id, not by name, so every operation needs
    /// the name→id mapping the last listing produced.
    private var fileIds: [String: String] = [:]

    init(auth: GoogleAuth) {
        self.auth = auth
    }

    // MARK: - RemoteFolder

    func list() async throws -> [RemoteFile] {
        let folder = try await folder()
        var items: [RemoteFile] = []
        var pageToken: String?

        repeat {
            var query = [
                URLQueryItem(name: "q", value: "'\(folder)' in parents and trashed=false"),
                URLQueryItem(name: "fields", value: "nextPageToken,files(id,name,modifiedTime,size,version)"),
                URLQueryItem(name: "pageSize", value: "200"),
            ]
            if let pageToken { query.append(URLQueryItem(name: "pageToken", value: pageToken)) }

            let listing: FileList = try await get("\(Self.api)/files", query: query)
            for file in listing.files {
                fileIds[file.name] = file.id
                items.append(RemoteFile(
                    name: file.name,
                    modifiedAt: ISO8601DateFormatter().date(from: file.modifiedTime ?? "") ?? .distantPast,
                    size: Int(file.size ?? "0") ?? 0,
                    // Drive bumps `version` on every content or metadata change,
                    // which is exactly the "has this changed" signal we need.
                    version: file.version ?? file.modifiedTime ?? "0"
                ))
            }
            pageToken = listing.nextPageToken
        } while pageToken != nil

        return items
    }

    func read(_ name: String) async throws -> Data {
        guard let id = try await id(for: name) else {
            throw RemoteFolderError.notFound(name)
        }
        return try await request(
            URLRequest(url: URL(string: "\(Self.api)/files/\(id)?alt=media")!)
        )
    }

    func write(_ name: String, data: Data) async throws {
        let token = try await auth.token()

        if let id = try await id(for: name) {
            // Existing file: replace its contents, keeping the same id so other
            // devices' bookmarks and the Drive revision history stay intact.
            var request = URLRequest(url: URL(string: "\(Self.upload)/files/\(id)?uploadType=media")!)
            request.httpMethod = "PATCH"
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = data
            _ = try await send(request)
            return
        }

        // New file: metadata and content in one multipart request.
        let folder = try await folder()
        let boundary = "mynote-\(UUID().uuidString)"
        let metadata = try JSONEncoder().encode(NewFile(name: name, parents: [folder]))

        var body = Data()
        body.append("--\(boundary)\r\nContent-Type: application/json; charset=UTF-8\r\n\r\n".data(using: .utf8)!)
        body.append(metadata)
        body.append("\r\n--\(boundary)\r\nContent-Type: application/json\r\n\r\n".data(using: .utf8)!)
        body.append(data)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)

        var request = URLRequest(url: URL(string: "\(Self.upload)/files?uploadType=multipart&fields=id")!)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("multipart/related; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        request.httpBody = body

        let created = try JSONDecoder().decode(DriveFile.self, from: try await send(request))
        fileIds[name] = created.id
    }

    func delete(_ name: String) async throws {
        guard let id = try await id(for: name) else { return }
        var request = URLRequest(url: URL(string: "\(Self.api)/files/\(id)")!)
        request.httpMethod = "DELETE"
        _ = try await send(request, authorize: true)
        fileIds.removeValue(forKey: name)
    }

    // MARK: - Folder resolution

    /// Find the MyNote folder, creating it on first use.
    private func folder() async throws -> String {
        if let folderId { return folderId }

        let query = [
            URLQueryItem(
                name: "q",
                value: "mimeType='\(Self.folderMime)' and name='\(Self.folderName)' and trashed=false"
            ),
            URLQueryItem(name: "fields", value: "files(id,name)"),
        ]
        let existing: FileList = try await get("\(Self.api)/files", query: query)
        if let found = existing.files.first {
            folderId = found.id
            return found.id
        }

        var request = URLRequest(url: URL(string: "\(Self.api)/files?fields=id")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(
            NewFile(name: Self.folderName, parents: nil, mimeType: Self.folderMime)
        )
        let created = try JSONDecoder().decode(DriveFile.self, from: try await send(request, authorize: true))
        folderId = created.id
        return created.id
    }

    private func id(for name: String) async throws -> String? {
        if let cached = fileIds[name] { return cached }
        _ = try await list()
        return fileIds[name]
    }

    // MARK: - Transport

    private func get<T: Decodable>(_ path: String, query: [URLQueryItem]) async throws -> T {
        var components = URLComponents(string: path)!
        components.queryItems = query
        let data = try await request(URLRequest(url: components.url!))
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func request(_ base: URLRequest) async throws -> Data {
        try await send(base, authorize: true)
    }

    private func send(_ base: URLRequest, authorize: Bool = false) async throws -> Data {
        var request = base
        if authorize || request.value(forHTTPHeaderField: "Authorization") == nil {
            request.setValue("Bearer \(try await auth.token())", forHTTPHeaderField: "Authorization")
        }

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw RemoteFolderError.offline
        }

        guard let http = response as? HTTPURLResponse else { throw RemoteFolderError.offline }
        guard http.statusCode < 400 else { throw Self.error(status: http.statusCode, body: data) }
        return data
    }

    private static func error(status: Int, body: Data) -> RemoteFolderError {
        let message = (try? JSONDecoder().decode(DriveError.self, from: body))?.error.message ?? "Drive error"
        switch status {
        case 401, 403 where message.localizedCaseInsensitiveContains("insufficient"):
            return .needsReauthentication
        case 403 where message.localizedCaseInsensitiveContains("quota"):
            return .storageFull
        case 404:
            return .notFound(message)
        case 429, 500...599:
            // Rate limited or Drive is unwell; both are worth retrying later.
            return .offline
        default:
            return .provider(message)
        }
    }

    // MARK: - Wire types

    private struct FileList: Decodable {
        let files: [DriveFile]
        let nextPageToken: String?
    }

    private struct DriveFile: Decodable {
        let id: String
        let name: String
        let modifiedTime: String?
        let size: String?
        let version: String?
    }

    private struct NewFile: Encodable {
        let name: String
        let parents: [String]?
        var mimeType: String?
    }

    private struct DriveError: Decodable {
        struct Inner: Decodable { let message: String }
        let error: Inner
    }
}
